### ISSUE TITLE
integer units causes a crash

### ISSUE THREAD

**Issue by JonathanGregory** [NCAS-CMS/cf-python#156]
On the RACC

    Platform: Linux-3.10.0-1062.18.1.el7.x86_64-x86_64-with-centos-7.7.1908-Core
    HDF5 library: 1.10.4
    netcdf library: 4.6.3
    udunits2 library: libudunits2.so.0
    python: 3.7.7
    netCDF4: 1.5.3
    cftime: 1.2.1
    numpy: 1.19.2
    psutil: 5.6.3
    scipy: 1.4.1
    matplotlib: 3.3.2
    ESMF: 7.1.0r
    cfdm: 1.8.7.1
    cfunits: 3.3.0
    cfplot: 3.0.32
    cf: 3.7.0
    >>> cf
    <module 'cf' from '/home/users/pg902583/jonathan/cf/__init__.py'>

    >>> r=cf.read('r.nc')
    Traceback (most recent call last):
    ....
    File "/home/users/pg902583/jonathan/cfdm/data/data.py", line 411, in __str__isreftime = ('since' in units)
    TypeError: argument of type 'numpy.int64' is not iterable

and

    netcdf r {
    dimensions:
        component_realization = 1000 ;
    variables:
        int64 component_realization(component_realization) ;
            component_realization:standard_name = "component_realization" ;
    double data(component_realization) ;
        data :units = 1LL ;
    // global attributes:
        :Conventions = "CF-1.8" ;
    }

This is my fault, because I had set `r.set_property('units',1)` i.e. integer 1 rather than string '1' but when I last ran this code a few months ago it did not cause a problem so I hadn't noticed. In fact this time I first hit the problem when I tried to multiply `r` by something else, which again ended with the conclusion `TypeError: argument of type 'numpy.int64' is not iterable`.

Although it's my fault, it's rather rude for it to crash, instead of trapping the error, I would say. Do you think it could be made more tolerant? Even more tolerantly, could you consider converting numeric units into a string, and only give an error if this is not possible?

Thanks for your help

Jonathan

**Comment by davidhassell** [comment]
Dear Jonathan,

I agree that the `TypeError: argument of type 'numpy.int64' is not iterable` should be handled better. It should not raise an exception at all, and it will be an easy fix.

I'm not sure why it got trapped from the read function - nor why it worked only a few months ago - could you possibly post the full traceback? This particular error is actually coming from the cfdm library, which does not use the `Units` object.

The error arising from the multiplication *is* a cf-python issue, and we can and will  handle that better, too.

I'm not so keen on converting non-string-valued input units to strings. This is because, to prevent failures when reading files with bad units, the `Units` object has been designed to accept anything, but record whether or not it is a valid unit:

```python
In [5]: import cf

In [6]: u = cf.Units(1)  # int

In [7]: u.isvalid
Out[7]: False

In [8]: u = cf.Units('1')  # str

In [9]: u.isvalid
Out[9]: True

In [10]: u = cf.Units('not a unit')  # bad str

In [11]: u.isvalid
Out[11]: False

```
Thanks,
David

**Comment by JonathanGregory** [comment]
Dear David

Thanks for investigating and for the explanation. Here's the traceback from the read error.

Best wishes

Jonathan

    >>> r=cf.read('r.nc')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/home/users/pg902583/jonathan/cfdm/decorators.py", line 186, in verbose_override_wrapper
        return method_with_verbose_kwarg(self, *args, **kwargs)
      File "/home/users/pg902583/jonathan/cf/read_write/read.py", line 656, in read
        warn_valid=warn_valid,
      File "/home/users/pg902583/jonathan/cfdm/decorators.py", line 186, in verbose_override_wrapper
        return method_with_verbose_kwarg(self, *args, **kwargs)
      File "/home/users/pg902583/jonathan/cf/read_write/read.py", line 903, in _read_a_file
        mask=mask, warn_valid=warn_valid)
      File "/home/users/pg902583/jonathan/cfdm/decorators.py", line 186, in verbose_override_wrapper
        return method_with_verbose_kwarg(self, *args, **kwargs)
      File "/home/users/pg902583/jonathan/cfdm/read_write/netcdf/netcdfread.py", line 1355, in read
        all_fields[ncvar] = self._create_field(ncvar)
      File "/home/users/pg902583/jonathan/cfdm/read_write/netcdf/netcdfread.py", line 2949, in _create_field
        "        [d] Inserting {!r}".format(data)
      File "/home/users/pg902583/jonathan/cf/data/data.py", line 1468, in __repr__
        return super().__repr__().replace('<', '<CF ', 1)
      File "/home/users/pg902583/jonathan/cfdm/data/data.py", line 212, in __repr__
        self.__class__.__name__, shape, str(self))
      File "/home/users/pg902583/jonathan/cfdm/data/data.py", line 411, in __str__
        isreftime = ('since' in units)
    TypeError: argument of type 'numpy.int64' is not iterable

**Comment by davidhassell** [comment]
This has, I hope, been resolved by cfunits v3.3.1 (https://github.com/NCAS-CMS/cfunits/issues/19) and cfdm v1.8.8.0 (https://github.com/NCAS-CMS/cfdm/commit/e3fb5ebdd776a6926f8887116649cacf753c9950)
